File: 0322-coin-change/0322-coin-change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        vector<int> dp(amount+1, 10005);
        int i;
        dp[0] = 0;
        for(i = 1; i <= amount; i++) {
            for(int coin: coins) {
                if(coin <= i) {
                    dp[i] = min(dp[i], 1 + dp[i-coin]);
                }
            }
        }
        return ((dp[amount] > amount) ? -1: dp[amount]);
    }
};

File: 0322-coin-change/0322-coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        minCoins = [amount + 1] * (amount + 1)
        minCoins[0] = 0
        for i in range(min(coins), amount + 1):
            for denom in coins:
                if denom <= i:
                    minCoins[i] = min(minCoins[i], 1 + minCoins[i - denom])
        
        return -1 if minCoins[amount] > amount else minCoins[amount]